File: 0994-rotting-oranges/0994-rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        w, h = len(grid[0]), len(grid)
        q = []
        hasFreshOrange = False
        for i, row in enumerate(grid):
            for j, orange in enumerate(row):
                if orange == 2:
                    q.append((i, j))
                if orange == 1:
                    hasFreshOrange = True
        # [[1]]
        # edge case: no rotten orange
        if not q:
            if hasFreshOrange: 
                return -1
            return 0

        minutes = -1
        while q:
            sz = len(q)
            for _ in range(sz):
                i, j = q[0]
                q.pop(0)
                
                # out of boundary / empty
                if i + 1 < h and grid[i+1][j] == 1: 
                    q.append((i+1, j))
                    grid[i+1][j] = 2
                if i - 1 >= 0 and grid[i-1][j] == 1: 
                    q.append((i-1, j))
                    grid[i-1][j] = 2
                if j + 1 < w and grid[i][j+1] == 1: 
                    q.append((i, j+1))
                    grid[i][j+1] = 2
                if j - 1 >= 0 and grid[i][j-1] == 1: 
                    q.append((i, j-1))
                    grid[i][j-1] = 2
            minutes += 1
        
        # edge case: cannot rot all the orange
        for i, row in enumerate(grid):
            for j, orange in enumerate(row):
                if orange == 1:
                    return -1
        
        return minutes

                

File: 0994-rotting-oranges/0994-rotting-oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        queue<pair<int, int>> pq;
        // push 
        for(int i = 0; i < grid.size(); i++){
            for(int j = 0; j < grid[i].size(); j++){
                if(grid[i][j] == 2){
                    pq.push({i, j});
                }
            }
        }
        vector<vector<int>> dirc = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        int minutes = 0;
        bool flag = false;
        while(!pq.empty()){
            int level = pq.size();
            if(!flag){minutes--; flag = true;}
            // if(minutes == 0) minutes --;
            minutes++;
            for(int j = 0; j < level; j++){
                pair<int, int> pr;
                pr = pq.front();
                pq.pop();
                int x = pr.first, y = pr.second;
                cout << x << y << endl;
                for(int i = 0; i < 4; i++){
                    if(x+dirc[i][0] > -1 && x+dirc[i][0] < grid.size() &&
                        y+dirc[i][1] > -1 && y+dirc[i][1] < grid[0].size() &&
                        grid[x+dirc[i][0]][y+dirc[i][1]] == 1){
                        grid[x+dirc[i][0]][y+dirc[i][1]] = 2;
                        pq.push({x+dirc[i][0], y+dirc[i][1]});
                    }
                }
            }
        }

        for(int i = 0; i < grid.size(); i++){
            for(int j = 0; j < grid[i].size(); j++){
                if(grid[i][j] == 1) return -1;
            }
        }
        return minutes;
    }
};